Cannot compile with Visual Studio 2017 RC (solved)

I tried to compile YouCompleteMe using the `install.py` script, but CMake cannot find the MSVC compiler. I was using CMake 3.7, the lastest CMake release. It seems that it cannot detect the new compiler. 
```powershell
λ python .\install.py --clang-completer
Searching Python 3.5 libraries...
Found Python library: C:\Users\yeben\Anaconda3\libs\python35.lib
Found Python headers folder: C:\Users\yeben\Anaconda3\include
-- The C compiler identification is unknown
-- The CXX compiler identification is unknown
CMake Error at CMakeLists.txt:26 (project):
  No CMAKE_C_COMPILER could be found.



CMake Error at CMakeLists.txt:26 (project):
  No CMAKE_CXX_COMPILER could be found.



-- Configuring incomplete, errors occurred!
See also "C:/Users/yeben/AppData/Local/Temp/ycm_build_4gbmd4_v/CMakeFiles/CMakeOutput.log".
See also "C:/Users/yeben/AppData/Local/Temp/ycm_build_4gbmd4_v/CMakeFiles/CMakeError.log".
ERROR: the build failed.

NOTE: it is *highly* unlikely that this is a bug but rather
that this is a problem with the configuration of your system
or a missing dependency. Please carefully read CONTRIBUTING.md
and if you're sure that it is a bug, please raise an issue on the
issue tracker, including the entire output of this script
and the invocation line used to run it.
```
Then I use the CMake installed with VS 2017 instead, and modify the `ycmd/build.py`. 
```Python
# ycmd/build.py
def BuildYcmdLib( args ):
  build_dir = mkdtemp( prefix = 'ycm_build_' )

  try:
    # full_cmake_args = [ '-G', GetGenerator( args ) ]
    full_cmake_args = [ '-G', "Visual Studio 15 2017 Win64" ]  # hack
    full_cmake_args.extend( CustomPythonCmakeArgs() )
    full_cmake_args.extend( GetCmakeArgs( args ) )
    full_cmake_args.append( p.join( DIR_OF_THIS_SCRIPT, 'cpp' ) )
```

I re-compiled, CMake successfully detected MSVC, but Boost failed to complie. Because it considers MSVC 1910 as an unknown version.
```C++
// ycmd\cpp\BoostParts\boost\config\compiler\visualc.hpp
//
// last known and checked version is 19.00.23026 (VC++ 2015 RTM):
#if (_MSC_VER > 1900)
#  if defined(BOOST_ASSERT_CONFIG)
#     error "Unknown compiler version - please run the configure tests and report the results"
#  else
#     pragma message("Unknown compiler version - please run the configure tests and report the results")
#  endif
#endif
```
I changed 1900 to 1910 and finally solved it.

I don't know how actually the build script work and I just found a hacky way to solve it. I wish you guys update the build script, which can be used to compile ycmd on windows easier :)

Not English native speaker, sorry for bad grammar.

--- a/appveyor.yml
+++ b/appveyor.yml
@@ -7,25 +7,27 @@ environment:
   USE_CLANG_COMPLETER: true
   COVERAGE: true
   matrix:
-  # We dropped MSVC 11 support because it lacks features from C++11
-  # and C99.
-  - msvc: 12
+  # We dropped MSVC 11 support because it lacks features from C++11 and C99.
+  - APPVEYOR_BUILD_WORKER_IMAGE: Visual Studio 2013
     arch: 64
     python: 36
-  - msvc: 14
+  - APPVEYOR_BUILD_WORKER_IMAGE: Visual Studio 2015
+    arch: 64
+    python: 36
+  - APPVEYOR_BUILD_WORKER_IMAGE: Visual Studio 2017
     arch: 32
     python: 36
-  # We only test Python 2.7 with MSVC 14 on 64 bits.
-  - msvc: 14
+  # We only test Python 2.7 with MSVC 15 on 64-bit.
+  - APPVEYOR_BUILD_WORKER_IMAGE: Visual Studio 2017
     arch: 64
     python: 27
-  - msvc: 14
+  - APPVEYOR_BUILD_WORKER_IMAGE: Visual Studio 2017
     arch: 64
     python: 36
 install:
   - ci\appveyor\appveyor_install.bat
 build_script:
-  - python run_tests.py --msvc %MSVC%
+  - ci\appveyor\appveyor_build.bat
 after_build:
   - codecov
 # Disable automatic tests

--- a/build.py
+++ b/build.py
@@ -224,17 +224,9 @@ def CustomPythonCmakeArgs():
 
 def GetGenerator( args ):
   if OnWindows():
-    if args.msvc == 14:
-      generator = 'Visual Studio 14'
-    elif args.msvc == 12:
-      generator = 'Visual Studio 12'
-    else:
-      generator = 'Visual Studio 11'
-
-    if platform.architecture()[ 0 ] == '64bit':
-      generator = generator + ' Win64'
-    return generator
-
+    return 'Visual Studio {version}{arch}'.format(
+        version = args.msvc,
+        arch = ' Win64' if platform.architecture()[ 0 ] == '64bit' else '' )
   if PathToFirstExistingExecutable( ['ninja'] ):
     return 'Ninja'
   return 'Unix Makefiles'
@@ -256,8 +248,8 @@ def ParseArguments():
   parser.add_argument( '--system-boost', action = 'store_true',
                        help = 'Use the system boost instead of bundled one. '
                        'NOT RECOMMENDED OR SUPPORTED!')
-  parser.add_argument( '--msvc', type = int, choices = [ 11, 12, 14 ],
-                       default = 14, help = 'Choose the Microsoft Visual '
+  parser.add_argument( '--msvc', type = int, choices = [ 12, 14, 15 ],
+                       default = 15, help = 'Choose the Microsoft Visual '
                        'Studio version (default: %(default)s).' )
   parser.add_argument( '--tern-completer',
                        action = 'store_true',

--- a/ci/appveyor/appveyor_build.bat
+++ b/ci/appveyor/appveyor_build.bat
@@ -1,0 +1,10 @@
+set msvc=%APPVEYOR_BUILD_WORKER_IMAGE:~-4%
+if %msvc% == 2013 (
+  set msvc=12
+) else if %msvc% == 2015 (
+  set msvc=14
+) else if %msvc% == 2017 (
+  set msvc=15
+)
+
+python run_tests.py --msvc %msvc%

--- a/run_tests.py
+++ b/run_tests.py
@@ -117,9 +117,9 @@ def ParseArguments():
                         COMPLETERS.keys()) )
   parser.add_argument( '--skip-build', action = 'store_true',
                        help = 'Do not build ycmd before testing.' )
-  parser.add_argument( '--msvc', type = int, choices = [ 11, 12, 14 ],
+  parser.add_argument( '--msvc', type = int, choices = [ 12, 14, 15 ],
                        help = 'Choose the Microsoft Visual '
-                       'Studio version. (default: 14).' )
+                       'Studio version. (default: 15).' )
   parser.add_argument( '--coverage', action = 'store_true',
                        help = 'Enable coverage report (requires coverage pkg)' )
   parser.add_argument( '--no-flake8', action = 'store_true',

